BUG Show actual stacktrace for errors / exceptions

tractorcow opened the pull request:
@phptek Sentry had been only ever rendering stack traces for the error handling classes themselves; These are not so useful when trying to diagnose the actual issue. :)

Tested on both fatal errors, as well as exceptions.

Uses the work I did at https://github.com/Seldaek/monolog/pull/1080 to capture non-exception traces as well. :)

tractorcow commented on src/Handler/SentryMonologHandler.php:
```diff
@@ -66,53 +66,54 @@ public function getClient()
-            'formatted'  => $record['formatted'],
-            'channel'    => $record['channel'],
+    {
+        $record = array_merge($record, [
```
added array_merge here to prevent other $record keys getting lost forever.

tractorcow commented on src/Handler/SentryRavenHandler.php:
```diff
@@ -80,10 +80,10 @@ protected function write(array $record)
-        if (!empty($record['stack'])) {
-            $stack = (bool) $record['stack'];
+        if (!empty($record['stacktrace'])) {
+            $stack = $record['stacktrace'];
```
Use actual trace array instead of bool to prevent sentry re-generating trace incorrectly.

phptek replied:
I dislike multi-type parameters intensely. If `Raven::captureMessage()` wants to accept an empty param for its 4th arg, it should allow an empty array....but then it goes on to accept `$stack` as `null` in `Raven::capture()` so what can you do...

I don't recall specifically why I cast to a `bool`, but it could be related to Raven_Client:909:

```
...
        if ((!$stack && $this->auto_log_stacks) || $stack === true) {
            $stack = debug_backtrace();
...
```

tractorcow replied:
If you cast it as a bool, it forces the stacktrace to be generated internally.

If you pass an array, it uses that value as the actual stacktrace.

The problem is that monolog stores the stacktrace for later error reporting (register_shutdown_function), so it's useless to try to use `debug_backtrace` in the capture since you are no longer in the stack.

I think false is just `no stacktrace`.

phptek commented on src/Handler/SentryMonologHandler.php:
```diff
@@ -9,55 +9,55 @@
 
     /**
-     * @param  int   $level
-     * @param  bool  $bubble
```
Which linting tool / ruleset are you using? I rather like vertically aligned param descriptions.

tractorcow replied:
PHPStorm auto-format, sorry. I can undo if you like.

tractorcow replied:
in Editor > Code Style > PHP

I've turned this on. :)

![image](https://user-images.githubusercontent.com/936064/49549638-35ca0f00-f94e-11e8-91f6-89e74a85b85a.png)

tractorcow commented on src/Handler/SentryMonologHandler.php:
```diff
@@ -122,28 +126,28 @@ private function backtrace($record)
-     * 
-     * @param  Member $member
+     *
+     * @param  Member       $member
```
How is this @phptek ?

tractorcow commented:
@phptek I've done a linting review, fixed a few type hints, fixed a bug in RavenClient::getLevel(), and fixed the indentation based on your preference. How does it look now?

phptek commented:
Merged. Many thanks @tractorcow

tractorcow commented:
Yay, thanks to you!

Now I feel bad about letting my PRs sit around for months when you have a same day turnaround, haha.

phptek commented:
LOL. I have standards to maintain seeing as PRs taking too long was the reason I built phptek/sentry instead of pushing PRs to the incumbent's Sentry module.

tractorcow commented:
@phptek can we please have a new release to include this fix? thanks.
